### PR TITLE
Eai 68

### PR DESCRIPTION
**Pull request by ntlong1594**

### CODE REVIEW

File: src/application/controllers/api/v1/Districts.php [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        if(!isset($district['name'])){
+            $this->_throwBadRequest('District name is required field');
+        }
+        $cityId = $district['id_city'];
+        if(empty($this->cities_model->findCityById($cityId))){
+            $this->_throwBadRequest('City id "'.$cityId.'" is not found');
+        }
+    }
+
+    public function getByIdAndNameAndCity(){
```
Comment (chicanh) [inline src/application/controllers/api/v1/Districts.php]: as the method body, it does not get by name, so remove the 'name' from method name
-> getByIdAndCity

File: src/application/controllers/api/v1/Districts.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        try{
+            $id = $this->input->get('id');
+            $id_city = $this->input->get('city');
+            if(!isset($id) && !isset($id_city)){
+                $this->_throwBadRequest('Either district_id or id_city must be defined in request param field to find District');
+            }
+            $result = $this->districts_model->findDistrictBy($id, $id_city);
```
Comment (chicanh) [inline src/application/controllers/api/v1/Districts.php:78]: is there any specific reason to include includes city name & id in GET district API due to the model query 
if not, please fix the model or create new one for this API

File: src/application/config/routes.php [hunk 3]
```diff
@@ -91,6 +94,12 @@
 $route['api/v1/settings/(:any)']['put'] = 'api/v1/settings/put/$1';
 $route['api/v1/settings/(:any)']['delete'] = 'api/v1/settings/delete/$1';
 
+// City api
+$route['api/v1/cities/create']['post'] = 'api/v1/cities/post';
```
Comment (jasminecrazy) [inline src/application/config/routes.php]: Do you have any reason to define such url.
If not have any reason please change to /cities

File: src/application/controllers/api/v1/API_V1_Controller.php [hunk 4]
```diff
@@ -90,6 +90,8 @@ protected function _handleException(\Exception $exception)
         $this->output
             ->set_content_type('application/json')
             ->set_output(json_encode($error, JSON_UNESCAPED_SLASHES | JSON_PRETTY_PRINT));
+            
+        exit($this->output->final_output);
```
Comment (jasminecrazy) [inline src/application/controllers/api/v1/API_V1_Controller.php:94]: I think we don't want to touch on v1. @chicanh what do you think ?

File: src/application/controllers/api/v1/Cities.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        $this->parser = new \EA\Engine\Api\V1\Parsers\Cities;
+    }
+
+    public function get(){
+        try{
+            $cities = $this->cities_model->getAllCities();
+         
+            $response = new Response($cities);
+        
+            $response->encode($this->parser)->output();
```
Comment (jasminecrazy) [inline src/application/controllers/api/v1/Cities.php:35]: Please add pagination also

File: src/application/controllers/api/v1/Cities.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            $response = new Response($city);
+            $response->output();
+        }
+        catch (\Exception $exception)
+        {
+            exit($this->_handleException($exception));
+        }
+    }
+
+    public function getByIdAndName(){
```
Comment (chicanh) [inline src/application/controllers/api/v1/Cities.php]: I think we just need the following API:

1. Get by id only
2. Search by name only - future support 

But getting by id and name is not real case I can see, am I right ?

File: src/application/controllers/api/v1/Cities.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        try{
+            $id = $this->input->get('id');
```
Comment (jasminecrazy) [inline src/application/controllers/api/v1/Cities.php]: Why do we need to get city by name and id also?
I think just one of them is ok

File: src/application/controllers/api/v1/Cities.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            $response = new Response($result);
+            $response->singleEntry(TRUE)->output();
+        }catch(\Exception $exception){
+            exit($this->_handleException($exception));
+        }
+    }
+
+    public function delete($id){
+        try{
+            if(!isset($id)){
```
Comment (jasminecrazy) [inline src/application/controllers/api/v1/Cities.php]: It's a path variable, so no need to check

File: src/application/config/routes.php [hunk 9]
```diff
@@ -91,6 +94,24 @@
...
```
Comment (chicanh) [inline src/application/config/routes.php]: -> $route['api/v1/cities/']['post'] = 'api/v1/cities/post';

File: src/application/config/routes.php [hunk 10]
```diff
@@ -91,6 +94,24 @@
...
+$route['api/v1/cities/get/all']['get'] = 'api/v1/cities/get';
```
Comment (chicanh) [inline src/application/config/routes.php]: -> $route['api/v1/cities']['get'] = 'api/v1/cities';

File: src/application/config/routes.php [hunk 11]
```diff
@@ -91,6 +94,24 @@
...
+$route['api/v1/cities/get']['get'] = 'api/v1/cities/getByIdAndName';
```
Comment (chicanh) [inline src/application/config/routes.php]: $route['api/v1/cities/:id']['get'] = 'api/v1/cities/$1';

File: src/application/config/routes.php [hunk 12]
```diff
@@ -91,6 +94,24 @@
...
+$route['api/v1/cities/delete/']['delete'] = 'api/v1/cities/delete/$1';
```
Comment (chicanh) [inline src/application/config/routes.php]: $route['api/v1/cities/:id']['delete'] = 'api/v1/cities/delete/$1';

File: src/application/config/routes.php [hunk 13]
```diff
@@ -91,6 +94,24 @@
...
+
+//District API
+$route['api/v1/districts/create']['post'] = 'api/v1/districts/post';
+$route['api/v1/districts/get/all']['get'] = 'api/v1/districts/get';
+$route['api/v1/districts/get']['get'] = 'api/v1/districts/getByIdAndNameAndCity';
```
Comment (chicanh) [inline src/application/config/routes.php]: $route['api/v1/districts/']['get'] = 'api/v1/districts/getByCity';
city id field param is required always

File: src/application/config/routes.php [hunk 14]
```diff
@@ -91,6 +94,24 @@
...
 $route['api/v1/settings/(:any)']['delete'] = 'api/v1/settings/delete/$1';
 
+// City api
+$route['api/v1/cities/create']['post'] = 'api/v1/cities/post';
+$route['api/v1/cities/get/all']['get'] = 'api/v1/cities/get';
+$route['api/v1/cities/get']['get'] = 'api/v1/cities/getByIdAndName';
+$route['api/v1/cities/delete/']['delete'] = 'api/v1/cities/delete/$1';
+
+//District API
+$route['api/v1/districts/create']['post'] = 'api/v1/districts/post';
```
Comment (chicanh) [inline src/application/config/routes.php]: -> $route['api/v1/districts']['post'] = 'api/v1/districts/post';

File: src/application/config/routes.php [hunk 15]
```diff
@@ -91,6 +94,24 @@
...
+$route['api/v1/districts/get/all']['get'] = 'api/v1/districts/get';
```
Comment (chicanh) [inline src/application/config/routes.php]: $route['api/v1/districts/get']['get'] = 'api/v1/districts/get';

File: src/application/config/routes.php [hunk 16]
```diff
@@ -91,6 +94,24 @@
...
+$route['api/v1/districts/get']['get'] = 'api/v1/districts/getByIdAndNameAndCity';
+$route['api/v1/districts/delete/']['delete'] = 'api/v1/districts/delete/$1';
```
Comment (chicanh) [inline src/application/config/routes.php]: $route['api/v1/districts/']['delete'] = 'api/v1/districts/delete/$1';

File: src/application/config/routes.php [hunk 17]
```diff
@@ -91,6 +94,24 @@
...
+
+//WARD API
+$route['api/v1/wards/create']['post'] = 'api/v1/wards/post';
```
Comment (chicanh) [inline src/application/config/routes.php]: $route['api/v1/wards/']['post'] = 'api/v1/wards/post';

File: src/application/config/routes.php [hunk 18]
```diff
@@ -91,6 +94,24 @@
...
+$route['api/v1/wards/get/all']['get'] = 'api/v1/wards/get';
```
Comment (chicanh) [inline src/application/config/routes.php]: $route['api/v1/wards']['get'] = 'api/v1/wards/getByDistricts';

File: src/application/controllers/api/v1/API_V1_Controller.php (hunk 4 above)
Comment (chicanh) [reply src/application/controllers/api/v1/API_V1_Controller.php:94]: @ntlong1594 sure please move all to v3 as latest API version number

File: src/application/controllers/api/v1/Cities.php (hunk 8 above)
Comment (ntlong1594) [reply src/application/controllers/api/v1/Cities.php]: I already removed

File: src/application/controllers/api/v1/Cities.php (hunk 7 above)
Comment (ntlong1594) [reply src/application/controllers/api/v1/Cities.php]: you can choose either of them or both of them

File: src/application/controllers/api/v1/Cities.php (hunk 5 above)
Comment (ntlong1594) [reply src/application/controllers/api/v1/Cities.php:35]: pagination is out of scope now :'(

File: src/application/config/routes.php (hunk 3 above)
Comment (ntlong1594) [reply src/application/config/routes.php]: I already removed.

File: src/application/controllers/api/v1/Cities.php (hunk 7 above)
Comment (jasminecrazy) [reply src/application/controllers/api/v1/Cities.php]: ok

**Review by jasminecrazy — APPROVED** [review]

File: src/assets/sql/structure.sql [hunk 19]
```diff
@@ -339,4 +343,46 @@ ALTER TABLE `ea_appointments`
...
+    ON UPDATE CASCADE;
+
+
+ALTER TABLE `ea_users`
+    ADD COLUMN `city_id` INT(11),
+    ADD COLUMN `district_id` INT(11),
+    ADD COLUMN `ward_id` INT(11),
+    ADD CONSTRAINT `fk_city_id` FOREIGN KEY (`city_id`) REFERENCES `integrated_cities` (`id`),
+    ADD CONSTRAINT `fk_district_id` FOREIGN KEY (`district_id`) REFERENCES `integrated_district` (`id`),
+    ADD CONSTRAINT `fk_ward_id` FOREIGN KEY (`ward_id`) REFERENCES `integrated_wards` (`id`),
```
Comment (chicanh) [inline src/assets/sql/structure.sql:387]: remove redundant comma

**Pull request merged**